packages/core/contracts/financial-templates/common/FundingRateApplier.sol

```diff
@@ -77,13 +89,60 @@ abstract contract FundingRateApplier is Testable, Lockable {
      * @notice Constructs the FundingRateApplier contract. Called by child contracts.
      * @param _finderAddress Finder used to discover financial-product-related contracts.
      */
-    constructor(address _finderAddress) public {
-        finder = FinderInterface(_finderAddress);
+    constructor(
+        FixedPoint.Unsigned memory _fundingRateBondPercentage,
+        FixedPoint.Unsigned memory _fundingRateRewardRate,
+        bytes32 _fundingRateIdentifier,
+        address _collateralAddress,
+        address _finderAddress,
+        address _timerAddress
+    ) public FeePayer(_collateralAddress, _finderAddress, _timerAddress) {
+        fundingRate.updateTime = getCurrentTime();
+
+        // Seed the cumulative multiplier as 1, from which it will be scaled as funding rates are applied over time.
+        fundingRate.cumulativeMultiplier = FixedPoint.fromUnscaledUint(1);
         emergencyShutdownTimestamp = 0;
-        lastUpdateTime = getCurrentTime();
 
-        // Seed the initial funding rate in the cumulativeFundingRateMultiplier 1.
-        cumulativeFundingRateMultiplier = FixedPoint.fromUnscaledUint(1);
+        fundingRate.rewardRate = _fundingRateRewardRate;
+        fundingRate.identifier = _fundingRateIdentifier;
+        fundingRate.bondPercentage = _fundingRateBondPercentage;
+    }
+
+    function proposeNewRate(FixedPoint.Signed memory rate, uint256 timestamp)
+        external
+        nonReentrant()
+        updateFundingRate()
+        returns (FixedPoint.Unsigned memory totalBond)
+    {
+        require(fundingRate.proposalTime == 0, "Proposal in progress");
+
+        // Timestamp must be after the last funding rate update time, within the last 30 minutes, and it cannot be more
+        // than 90 seconds ahead of the block timestamp.
+        uint256 currentTime = getCurrentTime();
+        uint256 updateTime = fundingRate.updateTime;
+        require(
+            timestamp > updateTime && timestamp >= currentTime.sub(30 minutes) && timestamp <= currentTime.add(90),
+            "Invalid proposal time"
+        );
+
+        // Set the proposal time in order to allow this contract to track this request.
+        fundingRate.proposalTime = timestamp;
+
+        OptimisticOracle optimisticOracle = _getOptimisticOracle();
+
+        // Set up optmistic oracle.
+        bytes32 identifier = fundingRate.identifier;
+        optimisticOracle.requestPrice(identifier, timestamp, collateralCurrency, 0);
+        totalBond = FixedPoint.Unsigned(
+            optimisticOracle.setBond(identifier, timestamp, _pfc().mul(fundingRate.bondPercentage).rawValue)
+        );
+
+        // Pull bond from caller and send to optimistic oracle.
+        if (totalBond.isGreaterThan(0)) {
+            collateralCurrency.safeTransferFrom(msg.sender, address(this), totalBond.rawValue);
+            collateralCurrency.safeIncreaseAllowance(address(optimisticOracle), totalBond.rawValue);
```

Comment:
is there a good reason to not set a large approval at the start of contract deployment to remove the need for this here? clearly, if the `optimisticOracle` gets compromised then it could drain the perp, which is a good enough reason in and of its self but. I thought it worthwhile to discuss.

Reply:
It's a great question. The reason I chose to do it here is that we're pulling the optimistic oracle address from the finder, so if it were upgraded for some reason, that big initial approval would no longer apply, which would make this revert (and lock up indefinitely I think). So, we'd effectively need to at least check and conditionally set the approval here if we wanted to do it that way. I figured it was simpler just to do the approval on each transaction rather than checking past approval. 

Thoughts?

Reply:
I agree with @mrice32 its more expensive gas-wise but it is more safe in that no transaction should fail

Reply:
Ye, good enough reason! Other option could be large approval at deployment then check approval size before doing the transfer and increase if need be. 

This is somewhat complex for no reason so I'm happy to keep it how you've got it here